### ISSUE TITLE
One-way audio after kernelization is kicked-in

### ISSUE THREAD

**Issue by jakubkarolczyk** [sipwise/rtpengine#1005]
Hi there,

I've found very weird issue with loosing an audio on one call leg just after kernelization is kicked-in.
It's really weird, because it's happening only when codec on both sides is negotiated to PCMA and when call setup is initiated from one side, towards (KAMAILIO/RTPENGINE->SBC->PSTN). From the other side (PSTN->SBC->KAMAILIO/RTPENGINE) it's fine, with the same codec.

Could you please have a look at the attached logs to find out if this is a bug or so?

If more information is needed, just let me know.

Thank you,
Jakub

-----

rtpengine -v
Version: git-master-115b73e

-----

dmesg | grep RTP
[  747.335869] xt_RTPENGINE: loading out-of-tree module taints kernel.
[  747.336218] xt_RTPENGINE: module verification failed: signature and/or required key missing - tainting kernel
[  747.337574] Registering xt_RTPENGINE module - version git-master-115b73e

-----

cat /proc/rtpengine/0/list
local inet4 1.2.3.4:10000
       src inet4 10.13.12.11:10000
       dst inet4 10.100.101.102:32182
    stats:                    0 bytes,                    0 packets,                    0 errors
    SRTP decryption (incoming) parameters:
        cipher: AES-CM-128
        HMAC: HMAC-SHA1
            auth tag length: 10
    option: rtcp-mux
    option: dtls
    option: stun
    option: transcoding
local inet4 10.13.12.11:10000
       src inet4 1.2.3.4:10000
       dst inet4 1.2.3.4:3613
    stats:                37668 bytes,                  219 packets,                    0 errors
        RTP payload type   8:                37668 bytes,                  219 packets
    SRTP encryption (outgoing) parameters:
        cipher: AES-CM-128
        HMAC: HMAC-SHA1
            auth tag length: 10
    option: transcoding

[rtpengine-01.log](https://github.com/sipwise/rtpengine/files/4710690/rtpengine-01.log)

**Comment by rfuchs** [comment]
The output port of both `offer` and `answer` shows up as 10000 in that log. Is that a result of you incorrectly anonymising it, or is that what's actually in the log?
```
m=audio 10000 RTP/AVP 8 101 18
...
m=audio 10000 UDP/TLS/RTP/SAVPF 8 101
```

**Comment by jakubkarolczyk** [comment]
That is actually in the log, as I didn't anonymise ports, only IPs.
I've just double check that behaviour and it looks like port 10000 (which is the first from my set range of ports) is always the same output port for both offer and answer for the first call after restart of rtpengine deamon. The second call, which I've made has 10014 as the output port for offer and 10012 for answer.
However for both calls, I lost the audio after few seconds, once kernelization was kicked-in.

**Comment by rfuchs** [comment]
Post your config? I'm guessing it might actually be port 10000 on both legs, but on different interfaces.

**Comment by jakubkarolczyk** [comment]
That's right, it's on two different interfaces

**Comment by rfuchs** [comment]
Ok, then please post a log that uses different port numbers. Otherwise it's very hard to follow what happens where.

**Comment by jakubkarolczyk** [comment]
Please find it attached

[rtpengine-02.log](https://github.com/sipwise/rtpengine/files/4711299/rtpengine-02.log)

**Comment by rfuchs** [comment]
What happens when you remove the transcoding options (both for G.729 and Opus)?

**Comment by jakubkarolczyk** [comment]
When I remove transcoding options all is fine. What I can see is that according to the below and logs, transcoding engine is not engage in that call.
If at least one codec is added as transcoded one, then transcoding engine is engage for the entire call even if negotiated codec is acceptable and the same on both side.

local inet4 51.52.49.54:10000
       src inet4 10.213.72.24:10000
       dst inet4 10.211.195.251:32204
    stats:                36036 bytes,                  198 packets,                    0 errors
        RTP payload type   8:                36036 bytes,                  198 packets
    SRTP decryption (incoming) parameters:
        cipher: AES-CM-128
        HMAC: HMAC-SHA1
            auth tag length: 10
    option: rtcp-mux
    option: dtls
    option: stun
local inet4 10.213.72.24:10000
       src inet4 51.52.49.54:10000
       dst inet4 51.52.49.54:9187
    stats:                34056 bytes,                  198 packets,                    0 errors
        RTP payload type   8:                34056 bytes,                  198 packets
    SRTP encryption (outgoing) parameters:
        cipher: AES-CM-128
        HMAC: HMAC-SHA1
            auth tag length: 10

**Comment by rfuchs** [comment]
Can you try with 9c67f08 ? I'm not sure I'm able to fully reproduce this so I'm not completely confident in the fix.

**Comment by jakubkarolczyk** [comment]
Still the same issue. Please find fresh logs attached.
[rtpengine-03.log](https://github.com/sipwise/rtpengine/files/4718348/rtpengine-03.log)

**Comment by rfuchs** [comment]
Ok, this seems to fix the unnecessary PCMA to PCMA transcoding. Both directions seem to get kernelised now. What are the contents of /proc/rtpengine/0/list after that happens?

**Comment by jakubkarolczyk** [comment]
Please find new logs attached and contents of /proc/rtpengine/0/list below

local inet4 1.2.3.4:10074
       src inet4 10.13.12.11:10084
       dst inet4 10.100.101.102:32330
    stats:                76804 bytes,                  422 packets,                    0 errors
        RTP payload type   8:                76804 bytes,                  422 packets
    SRTP decryption (incoming) parameters:
        cipher: AES-CM-128
        HMAC: HMAC-SHA1
            auth tag length: 10
    option: rtcp-mux
    option: dtls
    option: stun
local inet4 10.13.12.11:10084
       src inet4 10.100.101.102:10074
       dst inet4 10.100.101.102:3446
    stats:                72584 bytes,                  422 packets,                    0 errors
        RTP payload type   8:                72584 bytes,                  422 packets
    SRTP encryption (outgoing) parameters:
        cipher: AES-CM-128
        HMAC: HMAC-SHA1
            auth tag length: 10
    option: transcoding
[rtpengine-04.log](https://github.com/sipwise/rtpengine/files/4722397/rtpengine-04.log)

**Comment by rfuchs** [comment]
Ok so it seems that the kernel module is correctly seeing the packets and forwarding them. Please confirm through a packet dump (Wireshark or similar) if this actually happens. There might be some problem with the networking.

**Comment by jakubkarolczyk** [comment]
OK. I will do and let you know, however please note, when transcoding is disabled, all is fine and the call participants are exactly the same. Just for my curious, why "option: transcoding" is still visible in /proc/rtpengine/0/list on outgoing (affected) leg?

**Comment by rfuchs** [comment]
Because G.729 was offered and so that payload could potentially be sent and would then have to be transcoded.

**Comment by jakubkarolczyk** [comment]
I have one finding, which I think should bring you closer to the root cause. After kernelization is kicked in, our client starts throwing RTP decryption errors. The same happens when we try to decrypt them manually - packets sent before kernelization is kicked-in can be successfully decrypted and packets sent after kernelization is kicked-in cannot be decrypted. Below the output from rtp_decoder:
RTP packets decoded: 158 (before kernelization)
Packet decode errors: 140 (after kernelization)

Looks like kernelization+transcoding tandem break something in encyption or so.

**Comment by jakubkarolczyk** [comment]
Hi Richard, Please find fresh logs and raw (encrypted) RTP pcap attached.
The affected media stream is srcport=10000 --> dstport=7358

local inet4 1.2.3.4:10000
       src inet4 10.13.12.11:10000
       dst inet4 10.100.101.102:36908
    stats:                54782 bytes,                  301 packets,                    0 errors
        RTP payload type   8:                54782 bytes,                  301 packets
    SRTP decryption (incoming) parameters:
        cipher: AES-CM-128
        HMAC: HMAC-SHA1
            auth tag length: 10
    option: rtcp-mux
    option: dtls
    option: stun
local inet4 10.13.12.11:10000
       src inet4 1.2.3.4:10000
       dst inet4 1.2.3.4:7358
    stats:                52460 bytes,                  305 packets,                    0 errors
        RTP payload type   8:                52460 bytes,                  305 packets
    SRTP encryption (outgoing) parameters:
        cipher: AES-CM-128
        HMAC: HMAC-SHA1
            auth tag length: 10
    option: transcoding

[rtp_dump.pcap.txt](https://github.com/sipwise/rtpengine/files/4730428/rtp_dump.pcap.txt)

[rtpengine-05.log](https://github.com/sipwise/rtpengine/files/4730421/rtpengine-05.log)

**Comment by rfuchs** [comment]
Ok, so this confirms that there's something wrong with the encryption, but unfortunately I don't have an explanation as to why that happens. What kernel version and platform is this on?

I've updated the branch to 237d8958 which includes some extra debug output in the /proc file, in particular the SRTP keys which you can then compare to the ones from the log file. Hopefully that will shed some light on what's going on here.

**Comment by jakubkarolczyk** [comment]
So this is on Red Hat Enterprise Linux Server release 7.5 (Maipo) kernel 3.10.0-862.11.6.el7.x86_64.
Please find the fresh logs attached.

local inet4 1.2.3.4:10000
       src inet4 10.13.12.11:10000
       dst inet4 10.100.101.102:37010
    stats:                50050 bytes,                  275 packets,                    0 errors
        RTP payload type   8:                50050 bytes,                  275 packets
  SSRC in: 4fd20f4c
    SRTP decryption (incoming) parameters:
        cipher: AES-CM-128
    master key: a04d5af1b9e5fb5cd54559798bbcc56e
   master salt: d3a9fd984e1b05140db22bd3f2cc
   session key: 2d34ece6c59c105547a5e51c4b85811e
  session salt: 33def4251512a8e64f5800aa17f0
  session auth: a1871bd3483f5e3c9678575e0fe2652361c45059
           ROC: 0
        HMAC: HMAC-SHA1
            auth tag length: 10
    option: rtcp-mux
    option: dtls
    option: stun
local inet4 10.13.12.11:10000
       src inet4 1.2.3.4:10000
       dst inet4 1.2.3.4:5968
    stats:                47472 bytes,                  276 packets,                    0 errors
        RTP payload type   8:                47472 bytes,                  276 packets
  SSRC in: eebfc710
 SSRC out: a9979534
    SRTP encryption (outgoing) parameters:
        cipher: AES-CM-128
    master key: 3d5748f545f13a7163cbb1a16dab33c3
   master salt: 4857601ab862a00ef1ec65efd4be
   session key: 66bbd216c7310ba420fb3578d648fa1d
  session salt: f7c7e1330e6a9343d643e1d292bb
  session auth: 3fa4aea76eeb70513640c9c05a467cea2cbef136
           ROC: 0
        HMAC: HMAC-SHA1
            auth tag length: 10
    option: transcoding
[rtp_dump.pcap.txt](https://github.com/sipwise/rtpengine/files/4735605/rtp_dump.pcap.txt)
[rtpengine-06.log](https://github.com/sipwise/rtpengine/files/4735606/rtpengine-06.log)

**Comment by rfuchs** [comment]
Aha, I think that finally explains. Try with c0c9ce2 (which still includes the extra debug info) and see if that fixes it.

**Comment by jakubkarolczyk** [comment]
That's great, it works now.
However when I've generated a call in the other direction, it looks like it's not kernelized for the same PCMA codec on both ends. Please have a look at logs attached.

local inet4 1.2.3.4:10000
       src inet4 10.13.12.11:10000
       dst inet4 10.101.102.103:59644
    stats:                60424 bytes,                  332 packets,                    0 errors
        RTP payload type   8:                60424 bytes,                  332 packets
  SSRC in: 6c58f616
 SSRC out: a14e4558
    SRTP decryption (incoming) parameters:
        cipher: AES-CM-128
    master key: 90ecb3b9e445671f7e6c39225c14e6a9
   master salt: 0017e46906caf6fec6fce80b20f7
   session key: 5ad6a5b1e5d8d4989b367ed86fc31448
  session salt: a7a2e9bc3b98c9c3d29bbf29455b
  session auth: d2752c730a534051d10d808ff9753fa6ee949724
           ROC: 0
        HMAC: HMAC-SHA1
            auth tag length: 10
    option: rtcp-mux
    option: dtls
    option: stun
    option: transcoding
local inet4 10.13.12.11:10000
       src inet4 1.2.3.4:10000
       dst inet4 1.2.3.4:3746
    stats:                    0 bytes,                    0 packets,                    0 errors
  SSRC in: 9c5cc910
 SSRC out: 5368126c
    SRTP encryption (outgoing) parameters:
        cipher: AES-CM-128
    master key: 83faea82525b8c3e87b302cfb42c2f31
   master salt: 039509ece142bf2a284dee10dc64
   session key: 93ad602a4ecea17a19ff92c888fd17e7
  session salt: 66ce2b49374296cfee6370e957e2
  session auth: e0ef83891ce76c5d86f139e2e2f5e843a7071004
           ROC: 0
        HMAC: HMAC-SHA1
            auth tag length: 10
    option: transcoding


[rtpengine-07.log](https://github.com/sipwise/rtpengine/files/4736827/rtpengine-07.log)

**Comment by rfuchs** [comment]
I've added another commit a1043613 which I think should fix this.

**Comment by jakubkarolczyk** [comment]
All good now. Thanks Richard for all your work on that and quick fix.
I understand you merge this branch to master, so I will be able to update the official release?

**Comment by rfuchs** [comment]
Yes, as well as backported to 8.4 and 8.3 branches.
